REGRESSION (STP): Video won't play on bilibili.com

jyavenard opened the pull request:
#### 55a286438b239e94c5a355a1f7510cfa7b4dd033
<pre>
REGRESSION (STP): Video won&apos;t play on bilibili.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=263196">https://bugs.webkit.org/show_bug.cgi?id=263196</a>
rdar://117020123

Reviewed by Chris Dumez.

bilibili injects DOM objects in the tree by calling:
```
var t = (new DOMParser).parseFromString(e, &quot;text/html&quot;);
return document.adoptNode(t.body.firstChild)
```

Per spec, when parsing HTML to create a document through DOMParser::parseFromString() [1]
you are to skip the whitespace tokens for the &quot;initial&quot; insertion mode [2].
The trailing spaces however must be kept [3]
This behaviour was regressed by 267202@main.

On this site, the observable behaviour was that the firstChild became a &quot;#text&quot;
node rather the expected first DIV one.

So we trim the leading whitespaces before calling the fast parser. Which
restore the original behaviour prior 267202@main

[1] <a href="https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#parse-html-from-a-string">https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#parse-html-from-a-string</a>
[2] <a href="https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode">https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode</a>
[3] <a href="https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-afterbody">https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-afterbody</a>

* LayoutTests/fast/dom/document-contentType-DOMParser-expected.txt:
* LayoutTests/fast/dom/document-contentType-DOMParser.html: Added tests.
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::tryFastParsingHTMLFragment):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.h: Change type to StringView
(WebCore::requires):
* Source/WebCore/xml/DOMParser.cpp:
(WebCore::DOMParser::parseFromString):

Canonical link: <a href="https://commits.webkit.org/269457@main">https://commits.webkit.org/269457@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fc9ca812bbf9bd1fe81d0dfa6a7cb39fcbd9f4c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22624 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/269 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23711 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24533 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20943 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22883 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1353 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23154 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21893 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22864 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/180 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19628 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25386 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20499 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26744 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20740 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24583 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/175 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18034 "Passed tests") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20302 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5386 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -464,6 +464,7 @@ class HTMLFastPathParser {
 
     template<typename ParentTag> void parseCompleteInput()
     {
+        skipWhile<isASCIIWhitespace>(m_parsingBuffer);
```
This can't be correct. This changes the behavior of element.innerHTML too. However, I have just verified that this:
```
document.body.innerHTML = "\n        <div class=\"bpx-player-container\"></div>";
document.body.firstChild;
-> #text "\n        "
```

In Firefox.

I haven't checked the expected behavior for DOMParser's parseFromString() yet but at least the behavior change for innerHTML (which also uses this fast parser) is incorrect.

cdumez replied:
The trimming likely needs to happen in parseFromString(), just before calling the fast parser, not in the fast parser, since this behavior is specific to parseFromString().

cdumez commented on LayoutTests/fast/dom/document-contentType-DOMParser.html:
```diff
@@ -80,6 +80,9 @@
 shouldThrow('new DOMParser().parseFromString(xmlContent, "text/XML").contentType', "'TypeError: Type error'");
 shouldThrow('new DOMParser().parseFromString(htmlContent, "TEXT/html").contentType', "'TypeError: Type error'");
 
+var htmlContentWithSpace = "\n        <div class=\"bpx-player-container\"></div>";
```
Please add a similar test case for setting element.innerHTML to validate that it doesn't trim whitespace.

cdumez commented on LayoutTests/fast/dom/document-contentType-DOMParser.html:
```diff
@@ -80,6 +80,13 @@
+var htmlContentWithSpace = "\n        <div class=\"bpx-player-container\"></div>";
+shouldBeEqualToString('new DOMParser().parseFromString(htmlContentWithSpace, "text/html").body.firstChild.nodeName', 'DIV');
+
+var body = document.createElement("body");
```
Please use a `<div>`. I don't think it uses the HTML fast parser if the parent is a `<body>` element.

cdumez commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -201,6 +201,13 @@ class HTMLFastPathParser {
     {
     }
 
+    HTMLFastPathParser(StringParsingBuffer<CharacterType>&& parsingBuffer, Document& document, ContainerNode& destinationParent)
```
You should not need to modify the fast parser at all in this patch.

jyavenard replied:
The fastParser only takes a span from the original string, by doing the trimming there we avoid the requirement to create a new substring. No version where the trimming is done before calling the parser will be as optimal.

cdumez replied:
Update tryFastParsingHTMLFragment() to take a `StringView` instead of a `const String`. It shouldn't impact the implementation of the fast parser and it means the call site can create a substring view, which is cheap.

cdumez commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -895,7 +895,7 @@ static bool tryFastParsingHTMLFragmentImpl(const std::span<const CharacterType>&
 }
 
-bool tryFastParsingHTMLFragment(const String& source, Document& document, ContainerNode& destinationParent, Element& contextElement, OptionSet<ParserContentPolicy> policy)
+bool tryFastParsingHTMLFragment(const StringView& source, Document& document, ContainerNode& destinationParent, Element& contextElement, OptionSet<ParserContentPolicy> policy)
```
Just `StringView`, not `const StringView&`

cdumez commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.h:
```diff
@@ -41,7 +41,15 @@ class Element;
 
-WEBCORE_EXPORT bool tryFastParsingHTMLFragment(const String& source, Document&, ContainerNode&, Element& contextElement, OptionSet<ParserContentPolicy>);
+template<typename T>
+concept HasStringMethods = requires(T string)
```
What is this?

cdumez reviewed (approved):
LGTM assuming the bots are happy

jyavenard committed:
REGRESSION (STP): Video won't play on bilibili.com
https://bugs.webkit.org/show_bug.cgi?id=263196
rdar://117020123

Reviewed by Chris Dumez.

bilibili injects DOM objects in the tree by calling:
```
var t = (new DOMParser).parseFromString(e, "text/html");
return document.adoptNode(t.body.firstChild)
```

Per spec, when parsing HTML to create a document through DOMParser::parseFromString() [1]
you are to skip the whitespace tokens for the "initial" insertion mode [2].
The trailing spaces however must be kept [3]
This behaviour was regressed by 267202@main.

On this site, the observable behaviour was that the firstChild became a "#text"
node rather the expected first DIV one.

So we trim the leading whitespaces before calling the fast parser. Which
restore the original behaviour prior 267202@main

[1] https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#parse-html-from-a-string
[2] https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode
[3] https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-afterbody

* LayoutTests/fast/dom/document-contentType-DOMParser-expected.txt:
* LayoutTests/fast/dom/document-contentType-DOMParser.html: Added tests.
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::tryFastParsingHTMLFragment):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.h: Change type to StringView
(WebCore::requires):
* Source/WebCore/xml/DOMParser.cpp:
(WebCore::DOMParser::parseFromString):

Canonical link: https://commits.webkit.org/269457@main
